pulsar-jms-filters/src/main/java/com/datastax/oss/pulsar/jms/selectors/JMSPublishFilters.java

```diff
@@ -0,0 +1,153 @@
+    if (!enabled) {
+      return;
+    }
+    if (publishContext.isMarkerMessage()
+        || publishContext.isChunked()
+        || publishContext.getNumberOfMessages() > 1) {
+      return;
+    }
+
+    MessageMetadata messageMetadata =
+        Commands.peekMessageMetadata(headersAndPayload, "jms-filter-on-publish", -1);
+    if (messageMetadata.hasNumMessagesInBatch()) {
+      return;
+    }
+    producer
+        .getTopic()
+        .getSubscriptions()
+        .forEach(
+            (name, subscription) -> {
+              if (!(subscription instanceof PersistentSubscription)) {
+                return;
+              }
+              Map<String, String> subscriptionProperties = subscription.getSubscriptionProperties();
+              if (!subscriptionProperties.containsKey("jms.selector")) {
+                return;
+              }
+              FilterContext filterContext = new FilterContext();
+              filterContext.setSubscription(subscription);
+              filterContext.setMsgMetadata(messageMetadata);
+              filterContext.setConsumer(null);
+              Entry entry = null; // we would need the Entry only in case of batch messages
+              EntryFilter.FilterResult filterResult = filter.filterEntry(entry, filterContext);
+              if (filterResult == EntryFilter.FilterResult.REJECT) {
+                String property = "filter-result-" + name + "@" + subscription.getTopicName();
+                publishContext.setProperty(property, filterResult);
+                publishContext.setProperty(JMS_FILTERED_PROPERTY, true);
+              }
+            });
+  }
+
+  @Override
+  public void messageProduced(
+      ServerCnx cnx,
+      Producer producer,
+      long startTimeNs,
+      long ledgerId,
+      long entryId,
+      Topic.PublishContext publishContext) {
+    if (!enabled || publishContext.getProperty(JMS_FILTERED_PROPERTY) == null) {
+      return;
+    }
+    producer
+        .getTopic()
+        .getSubscriptions()
+        .forEach(
+            (name, subscription) -> {
+              String property = "filter-result-" + name + "@" + subscription.getTopicName();
+              EntryFilter.FilterResult filterResult =
+                  (EntryFilter.FilterResult) publishContext.getProperty(property);
+              if (filterResult == EntryFilter.FilterResult.REJECT) {
+                if (log.isDebugEnabled()) {
+                  log.debug("Reject message {}:{} for subscription {}", ledgerId, entryId, name);
+                }
+                // ir is possible that calling this method in this thread may affect performance
+                // let's keep it simple for now, we can optimize it later
+                subscription.acknowledgeMessage(
+                    Collections.singletonList(new PositionImpl(ledgerId, entryId)),
+                    CommandAck.AckType.Individual,
+                    null);
+              }
```

Comment:
I'm pretty sure acknowledging the message should be fast. It might get slowed down by synchronization, but I would expect the pre-filtering to be the thing that is most disruptive.